### DISCUSSION TITLE
Cobbler (v3.2.2) automatic deployment of ESXi and configuration of IP address with ks script error

### DISCUSSION THREAD

**Discussion by wangxu186189** [cobbler/cobbler#3548] [Version specific questions] [answered]
question：
![1](https://github.com/cobbler/cobbler/assets/29671452/beef5e28-9cc1-444a-968d-8469a6cdc40d)
![2](https://github.com/cobbler/cobbler/assets/29671452/668e8503-1a1d-411f-aff0-c582140c40ba)
![3](https://github.com/cobbler/cobbler/assets/29671452/c0f26faa-fd8c-4537-a271-1878956963b4)

env:  
os: RockyLinux8.9 
soft :  cobbler 3.2.2  cobbler-web 3.2.2
des:
I successfully deployed the ESXi 7.0.3 system in bulk to 100 servers using Cobbler (v3.2.2), and the corresponding kickstart script content is as follows: 

```
# Accept the VMware End User License Agreement
vmaccepteula

# Set the root password for ESXI HOST
rootpw password01!

# Install on the first local disk available on machine
install --firstdisk --overwritevmfs

# Set the network to dhcp on the first network adapter
network --bootproto=dhcp --device=vmnic0

# Set keyboard type
keyboard "US Default"

# Reboot
reboot

# Use busybox interpreter
%firstboot --interpreter=busybox

# Enable ESXi Shell
vim-cmd hostsvc/enable_esx_shell
vim-cmd hostsvc/start_esx_shell

# Enable SSH
vim-cmd hostsvc/enable_ssh
vim-cmd hostsvc/start_ssh

# Suppress Shell warning
esxcli system settings advanced set -o /UserVars/SuppressShellWarning -i 1

# enable password login (SSH)
sed -i "s/PasswordAuthentication no/PasswordAuthentication yes/g" /etc/ssh/sshd_confg;

# Reboot to apply settings 
esxcli system shutdown reboot -d 15
```

But when I want to achieve automatic batch deployment of ESXi 7. X system, when I automatically configure the management IP address (for vmnic0) and vmk0 address based on the server's bmc IP address, an error is reported on the cobbler web interface. When trying to install on the PXE network, there are also corresponding errors. Please help everyone. The code for my kickstart script is as follows:

```
# Accept the VMware End User License Agreement
vmaccepteula

# Set the root password for ESXI HOST
rootpw password01!

# Install on the first local disk available on machine
install --firstdisk --overwritevmfs

# Set the network to dhcp on the first network adapter
network --bootproto=dhcp --device=vmnic0

# Set keyboard type
keyboard "US Default"

# Reboot
reboot


# Use busybox interpreter
%firstboot --interpreter=busybox

# Enable ESXi Shell
vim-cmd hostsvc/enable_esx_shell
vim-cmd hostsvc/start_esx_shell

# Enable SSH
vim-cmd hostsvc/enable_ssh
vim-cmd hostsvc/start_ssh

# Suppress Shell warning
esxcli system settings advanced set -o /UserVars/SuppressShellWarning -i 1

# enable password login (SSH)
sed -i "s/PasswordAuthentication no/PasswordAuthentication yes/g" /etc/ssh/sshd_confg;

#get BMC IP
bmc_ip=$(esxcli hardware ipmi bmc get | grep IPv4Address | awk '{print $2}')

#redefine BMC IP
bmc_arr1=$(echo $bmc_ip|awk -F. '{print $1}')
bmc_arr2=$(echo $bmc_ip|awk -F. '{print $2}')
bmc_arr3=$(echo $bmc_ip|awk -F. '{print $3}')
bmc_arr4=$(echo $bmc_ip|awk -F. '{print $4}')

#define dw_arr3
dw_arr3=$(expr $bmc_arr3 - 100)

#define dw ip
dw_ip="$bmc_arr1.$bmc_arr2.$dw_arr3.$bmc_arr4"
dw_gw="$bmc_arr1.$bmc_arr2.$dw_arr3.254"

#setup dw ip
esxcli network ip interface ipv4 set -t static -I $dw_ip -N 255.255.255.0 -g $dw_gw -i vmk0

# Reboot to apply settings
esxcli system shutdown reboot -d 15
```

On the cobbler web interface, after clicking on the View Automatic Installation Template in my ESXI profiles, ks reported the following error:
```
# This automatic OS installation file had errors that prevented it from being rendered correctly.
# The cobbler.log should have information relating to this failure. 
```

[cobbler.log](https://github.com/cobbler/cobbler/files/13902430/cobbler.log)

I tried to directly use the ks script above to deploy ESXi 7. X system on a PXE network, but during the installation process, the DCUI interface reported the following error:
![1](https://github.com/cobbler/cobbler/assets/29671452/595d2390-c860-4df7-b048-2def9fc14005)
![2](https://github.com/cobbler/cobbler/assets/29671452/484d95f6-0a94-42a8-b55a-0e3fc528991b)
![3](https://github.com/cobbler/cobbler/assets/29671452/74fcd676-4042-4670-baee-5541594674d1)

Does Cobbler v3.2.2 support network configuration when automatically deploying ESXi OS with kickstart for ESXi 7. X and above?
Does anyone know how to solve this error?

**Comment by wangxu186189** [comment]
@SchoolGuy , Can u help me ？

**Reply by SchoolGuy** [reply]
You are missing the escaping of many of the characters that are special to Cheetah (the template language we use per default).

I recommend you read up the following page from their docs: https://cheetahtemplate.org/users_guide/language.html

The obvious thing that I spot for now is that all `$` need to be escaped with a backslash.

**Reply by wangxu186189** [reply]
many thanks , i will try again

**Reply by wangxu186189** [reply]
I have adjusted the code according to the document you provided, but there are still errors: 

```
# This automatic OS installation file had errors that prevented it from being rendered correctly.
# The cobbler.log should have information relating to this failure.
```

[cobbler.log](https://github.com/cobbler/cobbler/files/13919256/cobbler.log)

Here is my adjusted KS code :
```
# Accept the VMware End User License Agreement
vmaccepteula

# Set the root password for ESXI HOST
rootpw password01!

# Install on the first local disk available on machine
install --firstdisk --overwritevmfs

# Set the network to dhcp on the first network adapter
network --bootproto=dhcp --device=vmnic0

# Set keyboard type
keyboard "US Default"

# Reboot
reboot


# Use busybox interpreter
%firstboot --interpreter=busybox

# Enable ESXi Shell
vim-cmd hostsvc/enable_esx_shell \
vim-cmd hostsvc/start_esx_shell \

# Enable SSH
vim-cmd hostsvc/enable_ssh \
vim-cmd hostsvc/start_ssh  \

# Suppress Shell warning
esxcli system settings advanced set -o /UserVars/SuppressShellWarning -i 1 \

# enable password login (SSH)
sed -i "s/PasswordAuthentication no/PasswordAuthentication yes/g" /etc/ssh/sshd_confg; \

# get BMC IP
bmc_ip=$(esxcli hardware ipmi bmc get | grep IPv4Address | awk '{print $2}') \

#Splitting the BMC IP string to obtain various parts
bmc_arr1=$(echo $bmc_ip|awk -F. '{print $1}')    \
bmc_arr2=$(echo $bmc_ip|awk -F. '{print $2}')    \
bmc_arr3=$(echo $bmc_ip|awk -F. '{print $3}')    \
bmc_arr4=$(echo $bmc_ip|awk -F. '{print $4}')    \

# dw third byte
dw_arr3=$(expr $bmc_arr3 - 100) \

# dw IP address and gateway
dw_ip="$bmc_arr1.$bmc_arr2.$dw_arr3.$bmc_arr4" \
dw_gw="$bmc_arr1.$bmc_arr2.$dw_arr3.254" \

#setup dw ip
esxcli network ip interface ipv4 set -t static -I $dw_ip -N 255.255.255.0 -g $dw_gw -i vmk0 \

# Reboot to apply settings
esxcli system shutdown reboot -d 15 -r "Reboot to apply settings" \`
```

I adjusted my code according to the instructions in this section of the document :

Placeholders and directives can be escaped by putting a backslash before them. \$var and \#if will be output as literal text.
A placeholder or directive can span multiple physical lines, following the same rules as Python source code: put a backslash (\) at the end of all lines except the last line. However, if there’s an unclosed parenthesis, bracket or brace pending, you don’t need the backslash.

**Reply by wangxu186189** [reply]
@SchoolGuy  Is there a problem with my escape? I have attached the complete code, please take a look  
[test.zip](https://github.com/cobbler/cobbler/files/13919326/test.zip)

**Reply by SchoolGuy** [reply, marked as answer]
Try this one please:

```kickstart
# Accept the VMware End User License Agreement
vmaccepteula

# Set the root password for ESXI HOST
rootpw password01!

# Install on the first local disk available on machine
install --firstdisk --overwritevmfs

# Set the network to dhcp on the first network adapter
network --bootproto=dhcp --device=vmnic0

# Set keyboard type
keyboard "US Default"

# Reboot
reboot

# Use busybox interpreter
%firstboot --interpreter=busybox

# Enable ESXi Shell
vim-cmd hostsvc/enable_esx_shell \
vim-cmd hostsvc/start_esx_shell \

# Enable SSH
vim-cmd hostsvc/enable_ssh \
vim-cmd hostsvc/start_ssh  \

# Suppress Shell warning
esxcli system settings advanced set -o /UserVars/SuppressShellWarning -i 1 \

# enable password login (SSH)
sed -i "s/PasswordAuthentication no/PasswordAuthentication yes/g" /etc/ssh/sshd_confg; \

# get BMC IP
bmc_ip=\$(esxcli hardware ipmi bmc get | grep IPv4Address | awk '{print \$2}') \

#Splitting the BMC IP string to obtain various parts
bmc_arr1=\$(echo \$bmc_ip|awk -F. '{print \$1}')    \
bmc_arr2=\$(echo \$bmc_ip|awk -F. '{print \$2}')    \
bmc_arr3=\$(echo \$bmc_ip|awk -F. '{print \$3}')    \
bmc_arr4=\$(echo \$bmc_ip|awk -F. '{print \$4}')    \

# dw third byte
dw_arr3=\$(expr \$bmc_arr3 - 100) \

# dw IP address and gateway
dw_ip="\$bmc_arr1.\$bmc_arr2.\$dw_arr3.\$bmc_arr4" \
dw_gw="\$bmc_arr1.\$bmc_arr2.\$dw_arr3.254" \

#setup dw ip
esxcli network ip interface ipv4 set -t static -I \$dw_ip -N 255.255.255.0 -g \$dw_gw -i vmk0 \

# Reboot to apply settings
esxcli system shutdown reboot -d 15 -r "Reboot to apply settings" \`
```

I just escaped as previously described the `$` with a backslash.

**Reply by wangxu186189** [reply]
it works !  many thanks

**Reply by wangxu186189** [reply]
By the way, how can I view the logs on the cobbler during the PXE system installation process? Does the default cobbler not have a Print installation task log? Where can I configure it? I want to analyze the process of automatic deployment of the system in order to troubleshoot faults

**Reply by wangxu186189** [reply]
@SchoolGuy

**Reply by SchoolGuy** [reply]
@wangxu186189 all Cobbler logs are going to `/var/log/cobbler/cobbler.log`

The TFTP, DHCP and Apache services have their own respective logging files and folders.

The client installation logs are only available locally on the Cobbler server via the Anaconda installer when RH based systems are being installed. The client then sends those to Cobbler. Other vendors have not yet been implemented so far for this functionality.

**Reply by wangxu186189** [reply]
ok, maybe i can try to see the installation log by:  ctrl + alt + shift +F1 or   ctrl + alt + shift +F2

**Reply by wangxu186189** [reply]
I mean, when the pxe client encounters an installation error by pressing the combination key

**Reply by wangxu186189** [reply]
Although ESXi 7.0. X was successfully installed according to the new code you provided earlier, there were no further errors reported during the process. But when the installation task was completed and the server was automatically restarted, I found that the IP address and subnet mask of the management port vmnic0 were automatically set successfully, but the default gateway was empty, which is 0.0.0.0 or unset，as shown in the following figure： 

![image](https://github.com/cobbler/cobbler/assets/29671452/d801d7d5-4131-4f46-9035-f2a9b5fc6e46)
 
I looked through several logs in the ESXI7 system and found something strange:
/var/log/vmkernel. log: 
[vmkernel.log](https://github.com/cobbler/cobbler/files/13935854/vmkernel.log)

![error_in_vmkernel](https://github.com/cobbler/cobbler/assets/29671452/a0a4be9f-f89c-4263-85cd-4fce63772134)


/var/log/syslog.log : 
[syslog.log](https://github.com/cobbler/cobbler/files/13935861/syslog.log)

![error_in_syslog](https://github.com/cobbler/cobbler/assets/29671452/8259bd32-e9cd-4077-ac2c-423519ad8a29)

Additional explanation: The subnet where my cobbler server is located is 192.168.100.X/24. The subnet address that needs to be set for the pxe client (server) in the ks file is 181.33.101.X/24, and the gateway is 181.33.101.254. For details, please refer to the code logic in the ks configuration file you sent.
My question is, did the ks script execute successfully in its entirety? If the complete execution is successful, why is the default gateway for the vmnic0 network card of the ESXi host not set successfully? If the execution is not successful, is the problem still with the code logic in the ks file? Or it could be ESXI7 system

**Reply by wangxu186189** [reply]
@SchoolGuy Can you help me again ?

**Reply by SchoolGuy** [reply]
@wangxu186189 So what I can see is that your interface is provisioned via DHCP. As such check your DHCP server configuration. Is DHCP managed by Cobbler or externally?

**Reply by wangxu186189** [reply]
dhcp on cobbler ,here is my dhcp conf file:   
[esxi-zte.zip](https://github.com/cobbler/cobbler/files/13944191/esxi-zte.zip)

![image](https://github.com/cobbler/cobbler/assets/29671452/9b6c08a0-6479-4b2a-923e-49368d846d0c)
![image](https://github.com/cobbler/cobbler/assets/29671452/c882f401-4618-49ca-9205-a9dd5f8163d2)

the subnet network of cobbler is 192.168.100.0/24  , and the subnet of pxe client will be setup in different one.ext: 181.33.101.0/24
I guess this could be the reason for it

**Reply by SchoolGuy** [reply]
Yes this is indeed the reason. The DHCP server in the 181.33.101.0/24 network needs to hand out the corresponding gateway. If Cobbler is to manage the second network as well you need to add this network with its own subnet definition to the template Cobbler uses (`/etc/cobbler/dhcp.template`) to generate the `/etc/dhcpd.conf` file.

**Reply by wangxu186189** [reply]
ok, this is my current dhcp template in cobbler dir:   
[dhcp.zip](https://github.com/cobbler/cobbler/files/13952585/dhcp.zip)



Could you tell me How to reflect multiple subnets in this dhcp.template file ?  @SchoolGuy

**Reply by SchoolGuy** [reply]
@wangxu186189 RedHat has a nice help article for this: https://access.redhat.com/documentation/de-de/red_hat_enterprise_linux/7/html/networking_guide/sec-configuring_a_multihomed_dhcp_server
